lib/Shell.js

```diff
@@ -0,0 +1,239 @@
+const { EventEmitter } = require('events');
+const { spawn } = require('child_process');
+const os = require('os');
+const shortid = require('shortid');
+
+const PSCommand = require('./PSCommand');
+
+const {
+  isWin, isCriticalPSError, logger,
+  ShellStreamBuffer, shellSafeWrite,
+  getType, convertToPSOption,
+} = require('./utils');
+
+const {
+  PS_PROC_ERROR,
+  PS_ARG_MISS_ERROR,
+  PS_ARG_TYPE_ERROR,
+  PS_CMD_FAIL_ERROR,
+} = require('./errors');
+
+class Shell extends EventEmitter {
+  constructor(options = {}) {
+    super();
+
+    this.pid = -1;
+    this.streams = [];
+    this.commands = [];
+    this.history = [];
+    this.hadErrors = false;
+    this.invocationStateInfo = 'NotStarted';
+    this.verbose = false;
+
+    let psProc = !isWin ? 'pwsh' : 'powershell';
+    let psOpts = ['-NoLogo', '-NoExit', '-Command', '-'];
+
+    if(options.pwsh) {
+      psProc = 'pwsh';
+    }
+
+    if(options.pwshPrev && !options.pwsh) {
+      psProc = 'pwsh-preview';
+    }
+
+    if(process.env.NPS) {
+      psProc = process.env.NPS;
```

Comment:
what if both the env variable is set and options.pwsh/options.pwshPrev is passed?

Reply:
That's expected/documented?

Reply:
yap :)